### PR TITLE
Advanced contract tools

### PR DESCRIPTION
**Pull request by cyraxred**
* typed-by-name methods for contracts and events
* size analyzer
* storage layout analyzer

### CODE REVIEW

**Comment by cyraxred** [comment]
Will be replaced with https://github.com/windranger-io/hardhat-contract-tools

File: plugins/contract-sizes.ts [hunk 1]
```diff
@@ -0,0 +1,575 @@
...
+    HardhatConfig,
+    HardhatRuntimeEnvironment,
+    SolcConfig
+} from 'hardhat/types'
+import chalk from 'chalk'
+
+import fs from 'fs'
+import path from 'path'
+
+const MAX_CONTRACT_SIZE = 24576 // applied to color contract size as green / yellow / red
```
Comment (CjHare) [inline plugins/contract-sizes.ts:13]: suggestion: have this as file or env config with this as the default when absent.

Ideal contract sizing changes across networks. Specifically I'm thinking of the layer 2, that have support for much larger contract sizes.

File: plugins/contract-sizes.ts [hunk 2]
```diff
@@ -0,0 +1,575 @@
...
+    .addOptionalVariadicPositionalParam(
+        'contracts',
+        'Contracts to be printed, names or FQNs'
+    )
+    .setAction(async ({details: verbose, diff, size, contracts}, hre) => {
+        const c = (contracts ?? []) as string[]
+        const mappings = await extractBytecodeMappings(hre, c)
+        if (mappings.length === 0) {
+            // eslint-disable-next-line no-console
+            console.log('No contracts found')
```
Comment (CjHare) [inline plugins/contract-sizes.ts:55]: suggestion: use a logger over sop.

Bit more of a paint to setup, however it grants the user greater choice in what is done with the output

**Review by CjHare — COMMENTED** [review]
You're right, this would be best in it's own repo 👍

File: plugins/contract-sizes.ts (hunk 1 above)
Comment (cyraxred) [reply plugins/contract-sizes.ts]: this command is not related to any network, so
- filtering limits are provided via arguments
- and this one is only for coloring / cosmetics

ok, will make it as a CLI argument - making it as a per-network extension of hardhat config is an overkill

File: plugins/contract-sizes.ts (hunk 2 above)
Comment (cyraxred) [reply plugins/contract-sizes.ts]: afaik hardhat uses console, other hardhat plugins are using console ... why should this hardhat plugin be different?

File: package.json [hunk 3]
```diff
@@ -24,9 +24,10 @@
     "lint-sol": "solhint ./contracts/**/*.sol",
     "plant": "npx node-plantuml ./docs/specs",
     "prepare": "husky install",
+    "storage": "hardhat storage-layout --details",
     "sizes": "hardhat contract-sizes --details --diff --size 20000",
-    "sizes:all": "hardhat contract-sizes --diff",
-    "test": "mocha --timeout 10000 --exit --recursive --require ts-node/register \"test/**/*.test.ts\""
```
Comment (CjHare) [inline package.json:27]: question: are you dropping the test target?

Comment (cyraxred) [reply package.json:27]: no, it is the next line - I've just added a comma here

File: package.json [hunk 4]
```diff
@@ -54,6 +55,8 @@
     "ethereum-waffle": "3.4.4",
     "ethers": "5.6.9",
     "hardhat": "2.10.1",
+    "hardhat-contract-tools": "windranger-io/hardhat-contract-tools#2804de1752b6fb873f654fef033a8725fcfa1245",
```
[the remark this answers is not in the document]
Comment (cyraxred) [reply package.json]: it seems that for some reason you were looking on an wip commit
there are tags already

**Comment by CjHare** [comment]
Improvement in this PR being when the string given to `upgradgeContract` now cause a compile time error & mismatch of deploy args (inc type).

**Review by CjHare — APPROVED** [review]
👍

**Pull request merged**